flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java

```diff
@@ -241,6 +243,22 @@ public void putJobGraph(JobGraph jobGraph) throws Exception {
         LOG.info("Added {} to {}.", jobGraph, jobGraphStateHandleStore);
     }
 
+    @Override
+    public void putJobResourceRequirements(
+            JobID jobId, JobResourceRequirements jobResourceRequirements) throws Exception {
+        synchronized (lock) {
```

Comment:
Could you explain why do we need synchronization here?
It is already performed inside `recoverJobGraph` and `putJobGraph`. The latter also checks the version before writing, so there shouldn't be consistency issues IIUC.

Reply:
To prevent this from running concurrently with a `globalCleanupAsync`, which could result in a job graph being written after the cleanup.